fluentlenium-kotest/src/main/kotlin/org/fluentlenium/adapter/kotest/internal/KoTestFluentAdapter.kt

```diff
@@ -0,0 +1,120 @@
+package org.fluentlenium.adapter.kotest.internal
+
+import io.kotest.core.listeners.TestListener
+import io.kotest.core.spec.Spec
+import io.kotest.core.test.TestCase
+import io.kotest.core.test.TestResult
+import io.kotest.core.test.TestStatus
+import io.kotest.core.test.TestType
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import org.fluentlenium.adapter.DefaultSharedMutator
+import org.fluentlenium.adapter.FluentAdapter
+import org.fluentlenium.adapter.FluentTestRunnerAdapter
+import org.fluentlenium.adapter.IFluentAdapter
+import org.fluentlenium.adapter.TestRunnerCommon.*
+import org.fluentlenium.adapter.sharedwebdriver.SharedWebDriverContainer
+import org.fluentlenium.configuration.Configuration
+import org.fluentlenium.configuration.ConfigurationProperties
+import org.fluentlenium.utils.ScreenshotUtil
+import org.fluentlenium.utils.SeleniumVersionChecker
+import java.util.concurrent.atomic.AtomicReference
+
+internal class KoTestFluentAdapter constructor(var useConfigurationOverride: () -> Configuration = { throw IllegalStateException() }) :
+    IFluentAdapter,
+    FluentAdapter() {
+
+    val sharedMutator = DefaultSharedMutator()
+
+    /**
+     * KoTest doc states that tests inside a Spec are always executed sequentially (https://kotest.io/docs/framework/project-config.html#parallelism)
+     * So it should be safe to store the current test name like this.
+     */
+    val currentTestName = AtomicReference<String>()
+
+    private val configurationOverride: Configuration by lazy { useConfigurationOverride() }
+
+    override fun getConfiguration(): Configuration = configurationOverride
+
+    fun listener(): TestListener = object : TestListener {
+        override suspend fun beforeSpec(spec: Spec) {
+            require(driverLifecycle != ConfigurationProperties.DriverLifecycle.THREAD) {
+                // not sure about this:
+                // as the tests are executed within a Coroutine it
+                // could be possible that the "same" thread executes mulitple tests and/or that one test is executed
+                // by mulitple Coroutines with different underlying Threads. need to investigate
```

Comment:
This sounds related to [Configuration.kt#L125](https://github.com/kotest/kotest/blob/master/kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/config/Configuration.kt#L125)

Reply:
if i understand this correctly this means its safe to use one WebDriver per Thread if
`dispatcherAffinity == true`?

what is the best way to get the effective `dispatcherAffinity` value? there is a `var dispatcherAffinity: kotlin.Boolean?` in the spec but i am not sure if that is the effective one or only the override on spec level

Reply:
It's marked as kotest experimental, but you can get it like this `io.kotest.core.config.configuration.dispatcherAffinity` though I imagine that the spec level value is resolved from the project config. I can try to look into that tonight.

I might have given you a red herring though, so we should see what [Sam](https://github.com/sksamuel) thinks

Reply:
dispatcherAffinity == true means that the same thread is used for all tests in a spec. If that's true then yeah, you are single threaded. You could grab the value and fail if it's not set to true.

Most people are never going to change the value, and setting it true to use Fluentium seems fine, since interacting with the browser in a multi threaded way doesn't make sense anyway ?

The value can be overridden in the spec, so to get the value for a spec, you need to do:

`testCase.spec.dispatcherAffinity ?: testCase.spec.dispatcherAffinity() ?: configuration.dispatcherAffinity`